From 75f8370fc12364dcfafd3f97dc624e42650141c2 Mon Sep 17 00:00:00 2001
From: Arnold Csorvasi <arnold.csorvasi@ni.com>
Date: Tue, 8 Mar 2016 18:14:29 +0200
Subject: [PATCH] pkg_get_provider_replacees: do not add installed pkg to
 replacee list

If package A replaces provider B, and B is provided by A,
pkg_get_provider_replacees incorrectly adds A to the list of B replacees
when A is installed. During an upgrade, pacakge A is removed during
pkg_remove_installed_replacees, then once more during the package
upgrade.

Add check to skip the insertion of package A into the replacees vector
in pkg_get_provider_replacees.

Change taken and applied on v0.3.0 from commit 7885da3974de2df1baba220200d55920fb530e25.

Upstream-Status: Backport from 0.3.2

Signed-off-by: Arnold Csorvasi <arnold.csorvasi@ni.com>
---
 libopkg/opkg_install.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/libopkg/opkg_install.c b/libopkg/opkg_install.c
index dbfafa5..3b64b5f 100644
--- a/libopkg/opkg_install.c
+++ b/libopkg/opkg_install.c
@@ -427,10 +427,15 @@ static void pkg_get_provider_replacees(pkg_t * pkg,
             continue;
         for (j = 0; j < ap->pkgs->len; j++) {
             pkg_t *replacee = ap->pkgs->pkgs[j];
-            int installed = (replacee->state_status == SS_INSTALLED)
+            pkg_t *old = pkg_hash_fetch_installed_by_name(pkg->name);
+            /* skip pkg if installed: it  will be removed during upgrade
+               +             * issue 8913 */
+            if (old != replacee) {
+                int installed = (replacee->state_status == SS_INSTALLED)
                     || (replacee->state_status == SS_UNPACKED);
-            if (installed)
-                pkg_vec_insert(replacees, replacee);
+                if (installed)
+                    pkg_vec_insert(replacees, replacee);
+            }
         }
     }
 }
-- 
2.1.4

